File: tests/unit/core/dotnet/DearStory.Core.Tests/StorySessionTests.cs
using System.Text.Json.Nodes;
using DearStory.Core.Sessions;
using DearStory.Core.Services;
using Xunit;

namespace DearStory.Core.Tests;

public sealed class StorySessionTests
{
    [Fact]
    public void Open_creates_active_session_with_initial_state()
    {
        var defaultArguments = JsonNode.Parse("""{"label":"Save"}""")!;
        var session = StorySession.Open(
            Guid.Parse("11111111-1111-4111-8111-111111111111"),
            StoryId.Parse("buttons/primary"),
            defaultArguments,
            42,
            DateTimeOffset.Parse("2026-07-16T12:00:00Z"));

        Assert.Equal("buttons/primary", session.StoryId.Value);
        Assert.Equal("Save", session.Arguments["label"]!.GetValue<string>());
        Assert.False(session.IsClosed);
    }

    [Fact]
    public void Reset_restores_default_arguments_and_deterministic_services()
    {
        var defaultArguments = JsonNode.Parse("""{"label":"Save"}""")!;
        var session = StorySession.Open(
            Guid.Parse("11111111-1111-4111-8111-111111111111"),
            StoryId.Parse("buttons/primary"),
            JsonNode.Parse("""{"label":"Discard"}""")!,
            7,
            DateTimeOffset.Parse("2026-07-16T12:10:00Z"));

        session.Arguments["label"] = "Discard";
        session.Clock.Advance(TimeSpan.FromSeconds(30));
        _ = session.Random.NextUInt32();

        session.Reset(defaultArguments, 42, DateTimeOffset.Parse("2026-07-16T12:00:00Z"));

        Assert.Equal("Save", session.Arguments["label"]!.GetValue<string>());
        Assert.Equal(DateTimeOffset.Parse("2026-07-16T12:00:00Z"), session.Clock.CurrentUtc);
        Assert.Equal(new DeterministicRandom(42).NextUInt32(), session.Random.NextUInt32());
    }

    [Fact]
    public void Close_marks_session_closed()
    {
        var session = StorySession.Open(
            Guid.Parse("11111111-1111-4111-8111-111111111111"),
            StoryId.Parse("buttons/primary"),
            JsonNode.Parse("""{"label":"Save"}""")!,
            1,
            DateTimeOffset.Parse("2026-07-16T12:00:00Z"));

        session.Close("completed");

        Assert.True(session.IsClosed);
        Assert.Equal("completed", session.CloseReason);
    }
}

File: tests/unit/core/cpp/story_session_tests.cpp
#include <catch2/catch_test_macros.hpp>
#include <dearstory/core/story_session.hpp>

#include <chrono>
#include <nlohmann/json.hpp>

TEST_CASE("story_session opens with the requested initial state")
{
    auto const session = dearstory::core::story_session::open(
        "11111111-1111-4111-8111-111111111111",
        dearstory::core::story_id::parse("buttons/primary").value(),
        nlohmann::json::parse(R"({"label":"Save"})"),
        42U,
        std::chrono::sys_days{ std::chrono::July / 16 / 2026 } + std::chrono::hours{ 12 });

    REQUIRE(session.story().value() == "buttons/primary");
    REQUIRE(session.arguments().at("label") == "Save");
    REQUIRE_FALSE(session.is_closed());
}

TEST_CASE("story_session reset restores default arguments and deterministic services")
{
    auto session = dearstory::core::story_session::open(
        "11111111-1111-4111-8111-111111111111",
        dearstory::core::story_id::parse("buttons/primary").value(),
        nlohmann::json::parse(R"({"label":"Discard"})"),
        7U,
        std::chrono::sys_days{ std::chrono::July / 16 / 2026 } + std::chrono::hours{ 12 } + std::chrono::minutes{ 10 });

    session.arguments()["label"] = "Discard";
    session.clock().advance(std::chrono::seconds{ 30 });
    static_cast<void>(session.random().next_uint32());

    session.reset(
        nlohmann::json::parse(R"({"label":"Save"})"),
        42U,
        std::chrono::sys_days{ std::chrono::July / 16 / 2026 } + std::chrono::hours{ 12 });

    dearstory::core::deterministic_random expected_random{ 42U };

    REQUIRE(session.arguments().at("label") == "Save");
    REQUIRE(session.clock().current_utc() == (std::chrono::sys_days{ std::chrono::July / 16 / 2026 } + std::chrono::hours{ 12 }));
    REQUIRE(session.random().next_uint32() == expected_random.next_uint32());
    REQUIRE_FALSE(session.is_closed());
}

TEST_CASE("story_session close records a close reason")
{
    auto session = dearstory::core::story_session::open(
        "11111111-1111-4111-8111-111111111111",
        dearstory::core::story_id::parse("buttons/primary").value(),
        nlohmann::json::parse(R"({"label":"Save"})"),
        1U,
        std::chrono::sys_days{ std::chrono::July / 16 / 2026 } + std::chrono::hours{ 12 });

    session.close("completed");

    REQUIRE(session.is_closed());
    REQUIRE(session.close_reason() == "completed");
}
